File: algoexpert/easy/09.product_sum.ml
type item =
  | Value of int
  | Item of item list

let prod_sum lst =
  let rec f acc depth = function
    | Value (n) :: xs -> f (n+acc) depth xs
    | Item  (x) :: xs -> f acc (depth) xs + f 0 (depth+1) x
    | []              -> acc * depth
  in
  f 0 1 lst

let _ = prod_sum [
            Value(1);
            Value(2);
            Value(3);
            Item([Value(1);Value(3)]);
            Value(3);
          ]

File: algoexpert/easy/09.product_sum.py
#!/usr/bin/env python


def product_sum(lst, depth=1):
    sum = 0

    for el in lst:
        if type(el) is list:  # isinstance(el, type([]))
            sum += product_sum(el, depth+1)
        else:
            sum += el

    return sum * depth


if __name__ == '__main__':
    print(product_sum([1, 2, 3, [1, 3], 3]))
